### DISCUSSION TITLE
Help needed to run az acr build for Private ACR

### DISCUSSION THREAD

**Discussion by ghost** [Azure/ipam#321] [Q&A] [answered]
As suggested the steps
Next, use the following commands to update the Azure IPAM containers within your private Azure Container Registry

# App Services Container
az acr build -r <ACR Name> -t ipam:latest -f ./Dockerfile.deb .

I am using a Private endpoint connection for app service, Hence I am using Private ACR instead of pull the images from public container registry and using PE for container registry( I am hoping private access also can access to build images and pulled into Private ACR from the az acr build command)
Please suggest to run this command from self hosted agent pipeline from github actions. Any dependencies required?

**Comment by DCMattyG** [comment]
Hi @Vamsee-264, the great thing about using Docker is that everything is self-contained. The Dockerfile references the correct base images which have all of the dependencies included. We also use GitHub actions to build the production containers using the same commands.

Additionally, the `az acr build` command will package up all of the code and upload it to the Private ACR where it will be built, so you shouldn't need anything special to accomplish this in your GitHub Action runner outside of the Azure CLI available so the `az acr build` command can be exectued.

Hope that helps!

**Reply by ghost** [reply]
Hi @DCMattyG , Microsoft suggested this documentation - https://learn.microsoft.com/en-us/azure/container-registry/container-registry-private-link#execute-the-az-acr-build-with-private-endpoint-and-private-registry. So, want to understand it can be executed with or without dedicated acr agent pool?

I am using private endpoints to all services in my environment and disabled public access. (which is highly recommended)

The steps which followed:

Configuring resources using Terraform (with Private endpoints to all resources and web app service with vNet Integration). In order to provision using GitHub actions to build and deploy resources.
Using Azure ACR (not sure if public container build with az acr build and pushed into Private ACR).
Preparing to run the IPAM container service in Azure App service.

Let me know if the approach can fulfill to fully configured the IPAM or suggest a suitable solution please

**Reply by DCMattyG** [reply, marked as answer]
Hey @Vamsee-264, before I answer this in detail I'd just like to reiterate that at the end of the day, Azure IPAM runs in a container and uses a standard Dockerfile to be built. There are many ways to accomplish this, and many different places said image can be stored.

Your question to me looks more like an "Azure Support" question around how Private Container Registries work and the appropriate way to set them up in conjunction with GitHub Action runners. I would always advise to engage Azure support in questions like this as they are more qualified to speak to what can/cannot be done and assist with any troubleshooting.

That said, I did some testing for you, and here are my results...

Steps I followed:
- Created a Virtual Network where all of my Private Endpoints could live
- Created a test Virtual Machine and installed the Azure CLI to test executing the `az acr` command (simulating a GitHub runner)
- Created a Private Container Registry w/ Private Access using a Private Endpoint in a subnet within the Virtual Network
- I followed the directions in the document you linked above to create a Subnet for Agent Pool, and added the required Service Endpoints (as described [here](https://learn.microsoft.com/en-us/azure/container-registry/tasks-agent-pools#advanced-network-configuration))
- I created a NAT Gateway to allow traffic egress to the Internet for the Agent Pool VMSS (and attached it to the **pool** Subnet) as I'm not using any NVA's/Firewalls in my test setup (though you may very well be)
- Created a Subnet-integrated Agent Pool for my Private Container Registry using the Azure CLI command (shown [here](https://learn.microsoft.com/en-us/azure/container-registry/tasks-agent-pools#create-pool-in-vnet))
- Clone the Azure IPAM repo on my test Virtual Machine and ran the `az acr build` command in this format:
  - `az acr build -r <registry name> -t ipam:latest -f .\Dockerfile.deb . --agent-pool <agent pool name>`

With that, everything worked as expected.

I hope that helps. I'm happy to answer any further questions, but my conclusion at this point in time is that the document you shared above has all of the answers you need and by following it I was successfully able to build the container within a Private Container Registry w/ Private Endpoint and a Subnet-integrated Agent Pool.

**Reply by ghost** [reply]
Thank you for sharing the detailed technique @DCMattyG  and taking the time to for practical experiment. As described in the procedures, you constructed a NAT gateway to allow public internet traffic to access the public container image via the agent pool( could you please reference to do the same process as since Azure ACR has private endpoint( I understood why public access is needed, but wantedto understand technical design of it ). Also apologize for asking several questions, but I'd like to know the following for consistently upgrading the latest code. 
1. If we use a public container image, the application code is immediately downloaded and runs the newest code, but if I use ACR, do I need to pull the public container image into Private ACR and restart the app service each time? 
2. Is there automatic deployment for new releases in public container application code, or does Private ACR need to pull changes every time?

**Reply by DCMattyG** [reply]
Hey @Vamsee-264, allow me to respond to your questions...

> If we use a public container image, the application code is immediately downloaded and runs the newest code, but if I use ACR, do I need to pull the public container image into Private ACR and restart the app service each time?

Yes. You will be responsible for updating the images in the Private ACR. You can do this by building the images as we've discussed above, or copying the images from the Public ACR to the Private ACR by using the appropriate `az acr` commands.

> Is there automatic deployment for new releases in public container application code, or does Private ACR need to pull changes every time?

When using the Public ACR, your App Service will naturally restart from time to time as the instances move around in Azure, so you will get the new code eventually without much intervention. For Private ACR's you will need to manually update the images inside them.

Hope that helps to answer your questions, if not I'm here so feel free to keep 'em coming!

**Reply by ghost** [reply]
Hello @DCMattyG , 
I don't prefer private ACR because it always involves obtaining the latest container image whenever there's a new release( based on your confirmation) and having a large network set up, which incurs a more cost . My challenge here is to allow public container images from a public registry. 
 We are using the Private Endpoint and VNet interface with the app service (web app) to prevent public access and only allow resources within our VNet.
To allow public internet hosted code in the public registry, I guess we would  set up a NSG to allow public internet code.
If so, do we need to maintain a separate vNet for maintain all ipam resources privately( as we can't  prefer shared vnet to create  pe subnets) 
(it would be great if you could share a reference to allow public traffic to run ipam app in app service)?

**Reply by DCMattyG** [reply]
Do you not allow ANY public internet access?

How do you download updates (like Windows updates)?

Is there a firewall in place or something that controls the communication from the private network to the public Internet?

**Reply by ghost** [reply]
We generally use firewall policies to allow/deny public access, and for granular control over the internet using NSG, I would need to study how the setup is set up in my environment and will come with the details over the public access.

**Reply by DCMattyG** [reply]
Sounds good.

If you have a firewall, it would likely be that you would need to request a firewall rule to allow communication from the App Service to the publicly hosted ACR on port 443. Should be as simple as that.

Let me know what you find out!

**Reply by ghost** [reply]
Hi @DCMattyG ,
We are looking to enable a firewall rule to pull the image from public registry to private registry. Could you help us the details for creating a firewall rule to pull the image?

**Reply by DCMattyG** [reply]
Sure thing! What type of Firewall are we talking about here?

**Reply by ghost** [reply]
An Outbound firewall rule to allow outbound traffic to the Internet for the Agent Pool

**Reply by DCMattyG** [reply]
Yes, but is this like an Azure Firewall? An NVA like Palo Alto? Something else?

**Reply by ghost** [reply]
@DCMattyG  Yes It's Azure Firewall. Could you please provide details for configuration of Azure Firewall for Agent Pool to get/pull the public container image into Private ACR. 
https://github.com/Azure/ipam/discussions/321#discussioncomment-10584581 -- Thanks again for detailed  steps and taking a reference of these and need your help in Azure Firewall

**Reply by ghost** [reply]
Hi @DCMattyG, Could you please provide reference to allow access using Azure Firewall to pull the image into Private integrated ACR?

And please confirm Do we need to place Agent Pool dedicated subnet in github runner network setting to access it?

**Reply by DCMattyG** [reply]
@Vamsee-264, you will need an Agent Pool that lives within your Virtual Network so that it can access the Private ACR.

I spent some time doing testing, and here are my results...

Here is the Azure Firewall Policy I created:

![Screenshot from 2024-10-06 09-41-23](https://github.com/user-attachments/assets/7ad4fa29-bcb8-4984-91e8-ffee86e8f07a)

The first rule is to allow the Azure CLI to retrieve a token from Microsoft Login, and the second rule is to allow the Azure CLI to communicate to the Azure Management API.

From a Virtual Machine that is sitting inside my Virtual Network (to mimic an Agent Pool) I ran the following command to copy the container from the public Azure IPAM ACR into my private ACR, and then verified the results:

![Screenshot from 2024-10-06 09-48-37](https://github.com/user-attachments/assets/40758947-ac62-42d9-b4ad-94cc8fd37324)

For any additional items, your firewall team should be able to assist you as my above example is very specific to my environment and your configuration may vary. Here is a link to our documentation on configuring Firewall Rules for use with Azure Container Registry:

https://learn.microsoft.com/en-us/azure/container-registry/container-registry-firewall-access-rules

Hope that helped!

**Reply by ghost** [reply]
I sincerely appreciate the time and effort you dedicated to my query. Thank you so much! 
One last concern then will test this approach and  inform here with results🙂
Does this approach works when running az acr build ACR task which mentioned in the documentation https://azure.github.io/ipam/#/contributing/README?id=building-amp-updating-production-containers-images-using-a-private-acr

Or Can we mention destination container  FQDN something like " azureipam.azurecr.io "

**Reply by DCMattyG** [reply]
Hey @Vamsee-264, building the container from code and copying the container from a separate registry are two completely different things. You specifically mentioned that your ACR is privately integrated into your Virtual Network, and as such the above has nothing to do with `az acr build`. As long as your Agent Pool can reach your Private ACR (via being in the same Virtual Network or peering), then there shouldn't be any issues here.

Please reference my original response which already describes using `az acr build` with a Private ACR:

https://github.com/Azure/ipam/discussions/321#discussioncomment-10584581

**Reply by ghost** [reply]
Okay, that means there shouldn't be required any NAT gateway or Azure Firewall to be configured right. I was referencing the same comment and trying to understand why you've created NAT gateway for Agent Pool. If that shouldn't be required, that would be sufficient for us to test and proceed. Thanks!

**Reply by DCMattyG** [reply]
The NAT Gateway was to allow the VMSS Pool to reach out to the Internet to pull any required artifacts for the build process.

If your are using an Azure Firewall, you will need to allow access to those specific URL's if you don't allow default outbound access to HTTPS. The two I can think of off the top of my head are `github.com` so the VMSS Pool can pull down the code, and `docker.io` so the VMSS Pool can pull the base images from Docker Hub.

**Reply by DCMattyG** [reply]
Ok @Vamsee-264, I went through this one final time for you...

Here are the complete Firewall rules required to build the code in a Private ACR w/ Virtual Network integrated Agent Pool:

First, we need to allow the Service Tags which the VMSS will need to function.

![Screenshot from 2024-10-06 13-12-40](https://github.com/user-attachments/assets/beb217e1-14a0-4673-9e7a-f00f9fe95d75)

These are also documented [here](https://learn.microsoft.com/en-us/azure/container-registry/tasks-agent-pools#add-firewall-rules), and you have the option of using Service Endpoints instead, as I had mentioned in the original response above.

The we also need to allow all of the services/FQDNs required during the build process and those that are called from the Dockerfile.

![Screenshot from 2024-10-06 13-14-59](https://github.com/user-attachments/assets/30497aaa-3e40-457b-8735-9acc2ee756c0)

Based on my testing, this is just about the minimal set that should be required. You may have more broad firewall rules within your environment, but you should work with your Firewall team to determine the best way to implement these.

With that, I think we've exhausted every possible way to accomplish this, and I hope this thread has been helpful & informative.

**Reply by ghost** [reply]
@DCMattyG Thanks a ton! Sure, I will work with Networking team to determine if they are following any restrictive appliances w.r.t Firewalls. Meanwhile,  we will give a try to implement the possible ways which you mentioned 🙂. Thanks again for all your support!!!.
